In [2]:
# importing the necessary libraries
import pandas as pd
import numpy as np

In [3]:
df = pd.read_csv('data_interview_test.csv')

In [4]:
df.columns

Index(['receipt_id:company_id:matched_transaction_id:feature_transaction_id:DateMappingMatch:AmountMappingMatch:DescriptionMatch:DifferentPredictedTime:TimeMappingMatch:PredictedNameMatch:ShortNameMatch:DifferentPredictedDate:PredictedAmountMatch:PredictedTimeCloseMatch'], dtype='object')

In [5]:
df.shape

(12034, 1)

In [6]:
df.head(10)

receipt_id:company_id:matched_transaction_id:feature_transaction_id:DateMappingMatch:AmountMappingMatch:DescriptionMatch:DifferentPredictedTime:TimeMappingMatch:PredictedNameMatch:ShortNameMatch:DifferentPredictedDate:PredictedAmountMatch:PredictedTimeCloseMatch
10 000:10000:10 468:10  000:0.0000:0.0000:0.0000:1.0000:0.0000:0.0000:...                                                                                                                                                                                                                    
                468:10  001:0.0000:0.0000:0.0000:1.0000:0.0000:0.0000:...                                                                                                                                                                                                                    
                468:10  003:0.0000:0.0000:0.0000:1.0000:0.0000:0.0000:...                                                                                                                                                                                                                    
                468:10  004:0.0000:0.0000:0.0000:1.0000:0.0000:0.0000:...                                                                                                                                                                                                                    
                468:10  005:0.0000:0.0000:0.0000:1.0000:0.0000:0.0000:...                                                                                                                                                                                                                    
                468:10  006:0.0000:0.0000:0.0000:1.0000:0.0000:0.0000:...                                                                                                                                                                                                                    
                468:10  008:0.0000:0.0000:0.0000:1.0000:0.0000:0.0000:...                                                                                                                                                                                                                    
                468:10  009:0.0000:0.0000:0.0000:1.0000:0.0000:0.0000:...                                                                                                                                                                                                                    
                468:10  010:0.0000:0.0000:0.0000:1.0000:0.0000:0.0000:...                                                                                                                                                                                                                    
                468:10  011:0.0000:0.0000:0.0000:1.0000:0.0000:0.0000:...

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 12034 entries, (10, 000:10000:10, 468:10) to (50, 226:50000:50, 368:50)
Data columns (total 1 columns):
receipt_id:company_id:matched_transaction_id:feature_transaction_id:DateMappingMatch:AmountMappingMatch:DescriptionMatch:DifferentPredictedTime:TimeMappingMatch:PredictedNameMatch:ShortNameMatch:DifferentPredictedDate:PredictedAmountMatch:PredictedTimeCloseMatch    12034 non-null object
dtypes: object(1)
memory usage: 171.0+ KB


In [8]:
# Checking for the number of columns
for i in df.columns:
    print(len(i.split(':')))

14


In [9]:
# First splitting the values in the column based on ':'

for i in df.columns:
    column_names = i.split(':')


In [10]:
column_names

['receipt_id',
 'company_id',
 'matched_transaction_id',
 'feature_transaction_id',
 'DateMappingMatch',
 'AmountMappingMatch',
 'DescriptionMatch',
 'DifferentPredictedTime',
 'TimeMappingMatch',
 'PredictedNameMatch',
 'ShortNameMatch',
 'DifferentPredictedDate',
 'PredictedAmountMatch',
 'PredictedTimeCloseMatch']

In [11]:
# Splitting the values in rows on ':' 
df['list_col'] = df[df.columns[0]].str.split(':')

In [12]:
# Resetting the index to convert it from multiindex for easier accessibility
df = df.reset_index()

In [13]:
df.columns

Index(['level_0', 'level_1', 'level_2',
       'receipt_id:company_id:matched_transaction_id:feature_transaction_id:DateMappingMatch:AmountMappingMatch:DescriptionMatch:DifferentPredictedTime:TimeMappingMatch:PredictedNameMatch:ShortNameMatch:DifferentPredictedDate:PredictedAmountMatch:PredictedTimeCloseMatch',
       'list_col'],
      dtype='object')

In [14]:
column_names[3:14]

['feature_transaction_id',
 'DateMappingMatch',
 'AmountMappingMatch',
 'DescriptionMatch',
 'DifferentPredictedTime',
 'TimeMappingMatch',
 'PredictedNameMatch',
 'ShortNameMatch',
 'DifferentPredictedDate',
 'PredictedAmountMatch',
 'PredictedTimeCloseMatch']

In [15]:
# the first three columns should be renamed by the first three names in the column_names list 
# Splitting the list_col into new columns

df[column_names[3:14]] = pd.DataFrame(df.list_col.tolist(), index= df.index)

In [16]:
# Renaming the first three columns

df.rename(columns={'level_0':'receipt_id', 'level_1':'company_id', 'level_2':'matched_transaction_id'}, inplace=True)

In [17]:
df.columns

Index(['receipt_id', 'company_id', 'matched_transaction_id',
       'receipt_id:company_id:matched_transaction_id:feature_transaction_id:DateMappingMatch:AmountMappingMatch:DescriptionMatch:DifferentPredictedTime:TimeMappingMatch:PredictedNameMatch:ShortNameMatch:DifferentPredictedDate:PredictedAmountMatch:PredictedTimeCloseMatch',
       'list_col', 'feature_transaction_id', 'DateMappingMatch',
       'AmountMappingMatch', 'DescriptionMatch', 'DifferentPredictedTime',
       'TimeMappingMatch', 'PredictedNameMatch', 'ShortNameMatch',
       'DifferentPredictedDate', 'PredictedAmountMatch',
       'PredictedTimeCloseMatch'],
      dtype='object')

In [18]:
# Dropping off the columns which are already processed to create new columns

df = df[column_names]

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12034 entries, 0 to 12033
Data columns (total 14 columns):
receipt_id                 12034 non-null int64
company_id                 12034 non-null object
matched_transaction_id     12034 non-null object
feature_transaction_id     12034 non-null object
DateMappingMatch           12034 non-null object
AmountMappingMatch         12034 non-null object
DescriptionMatch           12034 non-null object
DifferentPredictedTime     12034 non-null object
TimeMappingMatch           12034 non-null object
PredictedNameMatch         12034 non-null object
ShortNameMatch             12034 non-null object
DifferentPredictedDate     12034 non-null object
PredictedAmountMatch       12034 non-null object
PredictedTimeCloseMatch    12034 non-null object
dtypes: int64(1), object(13)
memory usage: 1.3+ MB


In [20]:
# group by receipt id and member_id and create a list of all the elements in feature if matched in feature then, success else non-success

# Splitting the matched transaction id based on the ':'

df['rc_id_mtch_ftr_id'] = df['matched_transaction_id'].str.split(':')

In [21]:
# The 'rc_id_mtch_ftr_id' column contains the receipt id and feature transaction id

# First check if the receipt id matches and there're no anolmalies
# Then check for the transaction id using the logic:
# group by receipt id and member_id and create a list of all the elements in feature if matched in feature then, success else non-success


In [22]:
# check if the receipt id matches and there're no discrepancies

df['rc_id_mtch_ftr_id'][1]

['468', '10']

In [23]:
# The first value in the list contains the transaction id and the second element contains the receipt id
df['matched_receipt_id'] = df['rc_id_mtch_ftr_id'].apply(lambda x: x[1])


In [24]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12034 entries, 0 to 12033
Data columns (total 16 columns):
receipt_id                 12034 non-null int64
company_id                 12034 non-null object
matched_transaction_id     12034 non-null object
feature_transaction_id     12034 non-null object
DateMappingMatch           12034 non-null object
AmountMappingMatch         12034 non-null object
DescriptionMatch           12034 non-null object
DifferentPredictedTime     12034 non-null object
TimeMappingMatch           12034 non-null object
PredictedNameMatch         12034 non-null object
ShortNameMatch             12034 non-null object
DifferentPredictedDate     12034 non-null object
PredictedAmountMatch       12034 non-null object
PredictedTimeCloseMatch    12034 non-null object
rc_id_mtch_ftr_id          12034 non-null object
matched_receipt_id         12034 non-null object
dtypes: int64(1), object(15)
memory usage: 1.5+ MB


In [25]:
# Since, the matched receipt id is object, converting it to int

df['matched_receipt_id'] = df['matched_receipt_id'].astype(int)

In [26]:
df[df['receipt_id'] != df['matched_receipt_id']]

receipt_id    company_id matched_transaction_id feature_transaction_id  \
4741          30  002:30000:30                 592:31                    298   
4742          30  002:30000:30                 592:31                    299   
4743          30  002:30000:30                 592:31                    303   
4744          30  002:30000:30                 592:31                    310   
4769          30  004:30000:31                 109:31                    098   
4770          30  004:30000:31                 109:31                    099   
4771          30  004:30000:31                 109:31                    100   
4772          30  004:30000:31                 109:31                    101   
4773          30  004:30000:31                 109:31                    104   
4774          30  004:30000:31                 109:31                    105   
4775          30  004:30000:31                 109:31                    107   
4776          30  004:30000:31                 109:31                    108   
4777          30  004:30000:31                 109:31                    109   
4778          30  004:30000:31                 109:31                    110   
4784          30  006:30000:31                 446:31                    438   
4785          30  006:30000:31                 446:31                    441   
4786          30  006:30000:31                 446:31                    443   
4787          30  006:30000:31                 446:31                    444   
4788          30  006:30000:31                 446:31                    445   
4789          30  006:30000:31                 446:31                    446   
4790          30  006:30000:31                 446:31                    447   
4791          30  006:30000:31                 446:31                    448   
4792          30  007:30000:31                 056:31                    051   
4793          30  007:30000:31                 056:31                    052   
4794          30  007:30000:31                 056:31                    053   
4795          30  007:30000:31                 056:31                    054   
4796          30  007:30000:31                 056:31                    055   
4797          30  007:30000:31                 056:31                    056   
4798          30  007:30000:31                 056:31                    057   
4799          30  007:30000:31                 056:31                    058   
...          ...           ...                    ...                    ...   
9574          30  389:30000:31                 444:31                    448   
9590          30  393:30000:31                 463:31                    443   
9591          30  393:30000:31                 463:31                    444   
9592          30  393:30000:31                 463:31                    445   
9593          30  393:30000:31                 463:31                    446   
9594          30  393:30000:31                 463:31                    447   
9595          30  393:30000:31                 463:31                    448   
9596          30  393:30000:31                 463:31                    449   
9597          30  393:30000:31                 463:31                    450   
9598          30  393:30000:31                 463:31                    451   
9599          30  393:30000:31                 463:31                    452   
9600          30  393:30000:31                 463:31                    453   
9601          30  393:30000:31                 463:31                    454   
9602          30  393:30000:31                 463:31                    455   
9603          30  393:30000:31                 463:31                    456   
9604          30  393:30000:31                 463:31                    457   
9605          30  393:30000:31                 463:31                    458   
9606          30  393:30000:31                 463:31                    459   
9607  

#### There are 820 records where the receipt id extracted does not match with the exact receipt id

In [27]:
# Extracting the transaction_id from the matched_transaction_id column

df['mtchd_trnx_id'] = df['rc_id_mtch_ftr_id'].apply(lambda x: x[0])

In [28]:
df[(df['mtchd_trnx_id']==df['feature_transaction_id']) & (df['matched_receipt_id']==df['receipt_id'])]

receipt_id    company_id matched_transaction_id feature_transaction_id  \
25             10  001:10000:10                 605:10                    605   
27             10  002:10000:10                 286:10                    286   
38             10  003:10000:10                 412:10                    412   
73             10  005:10000:10                 474:10                    474   
105            10  007:10000:10                 360:10                    360   
138            10  009:10000:10                 484:10                    484   
153            10  010:10000:10                 405:10                    405   
170            10  011:10000:10                 659:10                    659   
173            10  012:10000:10                 256:10                    256   
207            10  014:10000:10                 650:10                    650   
210            10  015:10000:10                 713:10                    713   
232            10  017:10000:10                 504:10                    504   
244            10  018:10000:10                 400:10                    400   
255            10  019:10000:10                 649:10                    649   
262            10  020:10000:10                 683:10                    683   
267            10  021:10000:10                 112:10                    112   
278            10  022:10000:10                 404:10                    404   
281            10  023:10000:10                 255:10                    255   
302            10  024:10000:10                 332:10                    332   
312            10  025:10000:10                 324:10                    324   
317            10  026:10000:10                 508:10                    508   
351            10  028:10000:10                 279:10                    279   
356            10  029:10000:10                 574:10                    574   
361            10  030:10000:10                 676:10                    676   
375            10  031:10000:10                 561:10                    561   
385            10  032:10000:10                 302:10                    302   
388            10  033:10000:10                 140:10                    140   
402            10  034:10000:10                 556:10                    556   
408            10  035:10000:10                 371:10                    371   
410            10  036:10000:10                 444:10                    444   
...           ...           ...                    ...                    ...   
11782          50  188:50000:50                 137:50                    137   
11792          50  189:50000:50                 338:50                    338   
11795          50  190:50000:50                 247:50                    247   
11806          50  191:50000:50                 047:50                    047   
11810          50  192:50000:50                 067:50                    067   
11814          50  193:50000:50                 222:50                    222   
11818          50  194:50000:50                 096:50                    096   
11843          50  196:50000:50                 365:50                    365   
11849          50  197:50000:50                 167:50                    167   
11850          50  198:50000:50                 251:50                    251   
11854          50  199:50000:50                 075:50                    075   
11863          50  200:50000:50                 099:50                    099   
11870          50  201:50000:50                 100:50                    100   
11874          50  202:50000:50                 126:50                    126   
11877          50  203:50000:50                 280:50                    280   
11883          50  204:50000:50                 143:50                    143   
11888          50  205:50000:50                 150:50                    150   
11893          50  208:50000:50       

In [29]:
# There are 788 records which are success

# Creating a new column with a success value as 1 when the receipt id and the transaction id match else 0

df['Success'] = np.where((df['mtchd_trnx_id']==df['feature_transaction_id']) & (df['matched_receipt_id']==df['receipt_id']), 1, 0)

In [30]:
# Groupby receipt and feature_transaction_id, sort by Success value and then, drop the success column
df = df.groupby(['receipt_id', 'feature_transaction_id']).apply(lambda x: x.sort_values(by='Success', ascending=False)).drop('Success', axis=1)

In [31]:
df

receipt_id    company_id  \
receipt_id feature_transaction_id                                   
10         000                    0              10  000:10000:10   
                                  1448           10  135:10000:10   
                                  1510           10  141:10000:10   
                                  1530           10  143:10000:10   
                                  1550           10  144:10000:10   
                                  1570           10  145:10000:10   
                                  1629           10  150:10000:10   
                                  1710           10  157:10000:10   
                                  1752           10  161:10000:10   
                                  1781           10  164:10000:10   
                                  1834           10  167:10000:10   
                                  1865           10  170:10000:10   
                                  1885           10  171:10000:10   
                                  1926           10  175:10000:10   
                                  1973           10  180:10000:10   
                                  1993           10  182:10000:10   
                                  2056           10  188:10000:10   
                                  2106           10  195:10000:10   
                                  2129           10  197:10000:10   
                                  2167           10  200:10000:10   
                                  2200           10  203:10000:10   
                                  2235           10  209:10000:10   
                                  2255           10  210:10000:10   
                                  2265           10  211:10000:10   
                                  2285           10  212:10000:10   
                                  1470           10  137:10000:10   
                                  1382           10  130:10000:10   
                                  42             10  004:10000:10   
                                  1355           10  128:10000:10   
                                  77             10  006:10000:10   
...                                             ...           ...   
50         367                    11978          50  221:50000:50   
                                  12032          50  226:50000:50   
           368                    12033          50  226:50000:50   
                                  10678          50  009:50000:50   
                                  10876          50  044:50000:50   
                                  10970          50  058:50000:50   
                                  11857          50  199:50000:50   
           369                    10679          50  009:50000:50   
                                  10877          50  044:50000:50   
           370                    11577          50  157:50000:50   
           376                    11444          50  132:50000:50   
                                  11638          50  168:50000:50   
           377                    11445          50  132:50000:50   
                                  11639          50  168:50000:50   
           378                    10812          50  033:50000:50   
           379                    10813          50  033:50000:50   
           407                    11134          50  089:50000:50   
                                  11078          50  079:50000:50   
           408                    11079          50  079:50000:50   
                                  11135          50  089:50000:50   
           409                    11080          50  079:50000:50   
                                  11136          50  089:50000:50   
           410                    11081          50  079:50000:50   
                                  11137          50  089:50000:50   
           411                    11082          50  079:50000:50   
                                  11138

In [33]:
df.columns

17

In [36]:
# Dropping the unnecessary columns

df = df[column_names]

In [37]:
df

receipt_id    company_id  \
receipt_id feature_transaction_id                                   
10         000                    0              10  000:10000:10   
                                  1448           10  135:10000:10   
                                  1510           10  141:10000:10   
                                  1530           10  143:10000:10   
                                  1550           10  144:10000:10   
                                  1570           10  145:10000:10   
                                  1629           10  150:10000:10   
                                  1710           10  157:10000:10   
                                  1752           10  161:10000:10   
                                  1781           10  164:10000:10   
                                  1834           10  167:10000:10   
                                  1865           10  170:10000:10   
                                  1885           10  171:10000:10   
                                  1926           10  175:10000:10   
                                  1973           10  180:10000:10   
                                  1993           10  182:10000:10   
                                  2056           10  188:10000:10   
                                  2106           10  195:10000:10   
                                  2129           10  197:10000:10   
                                  2167           10  200:10000:10   
                                  2200           10  203:10000:10   
                                  2235           10  209:10000:10   
                                  2255           10  210:10000:10   
                                  2265           10  211:10000:10   
                                  2285           10  212:10000:10   
                                  1470           10  137:10000:10   
                                  1382           10  130:10000:10   
                                  42             10  004:10000:10   
                                  1355           10  128:10000:10   
                                  77             10  006:10000:10   
...                                             ...           ...   
50         367                    11978          50  221:50000:50   
                                  12032          50  226:50000:50   
           368                    12033          50  226:50000:50   
                                  10678          50  009:50000:50   
                                  10876          50  044:50000:50   
                                  10970          50  058:50000:50   
                                  11857          50  199:50000:50   
           369                    10679          50  009:50000:50   
                                  10877          50  044:50000:50   
           370                    11577          50  157:50000:50   
           376                    11444          50  132:50000:50   
                                  11638          50  168:50000:50   
           377                    11445          50  132:50000:50   
                                  11639          50  168:50000:50   
           378                    10812          50  033:50000:50   
           379                    10813          50  033:50000:50   
           407                    11134          50  089:50000:50   
                                  11078          50  079:50000:50   
           408                    11079          50  079:50000:50   
                                  11135          50  089:50000:50   
           409                    11080          50  079:50000:50   
                                  11136          50  089:50000:50   
           410                    11081          50  079:50000:50   
                                  11137          50  089:50000:50   
           411                    11082          50  079:50000:50   
                                  11138

In [39]:
# function for the model that takes in the file location and spits out a csv with Success matching at the top 

def preprocess(filelocation):
    """
    Input - filelocation, file type should be a csv
    Output - A dataframe with the correct transaction if exists at the top
    
    """
    # Converting the csv file to pandas dataframe
    df = pd.read_csv('data_interview_test.csv')
    
    # First splitting the column names on ':'

    for i in df.columns:
        column_names = i.split(':')
        
    # Splitting the values in rows on ':' 
    
    df['list_col'] = df[df.columns[0]].str.split(':')
    
    # Resetting the index to convert it from multiindex for easier accessibility
    
    df = df.reset_index()
    
    # the first three columns should be renamed by the first three names in the column_names list 
    # Splitting the list_col into new columns
    
    df[column_names[3:14]] = pd.DataFrame(df.list_col.tolist(), index= df.index)
    
    # Renaming the first three columns

    df.rename(columns={'level_0':'receipt_id', 'level_1':'company_id', 'level_2':'matched_transaction_id'}, inplace=True)
    
    # Dropping off the columns which are already processed to create new columns

    df = df[column_names]
    
    # Splitting the matched transaction id based on the ':'

    df['rc_id_mtch_ftr_id'] = df['matched_transaction_id'].str.split(':')
    
    # The first value in the list contains the transaction id and the second element contains the receipt id
    
    df['matched_receipt_id'] = df['rc_id_mtch_ftr_id'].apply(lambda x: x[1])
    
    # Since, the matched receipt id is object, converting it to int

    df['matched_receipt_id'] = df['matched_receipt_id'].astype(int)
    
    # Extracting the transaction_id from the matched_transaction_id column

    df['mtchd_trnx_id'] = df['rc_id_mtch_ftr_id'].apply(lambda x: x[0])

    # Creating a new column with a success value as 1 when the receipt id and the transaction id match else 0

    df['Success'] = np.where((df['mtchd_trnx_id']==df['feature_transaction_id']) & (df['matched_receipt_id']==df['receipt_id']), 1, 0)
    
    # Groupby receipt and feature_transaction_id, sort by Success value and then, drop the success column
    
    df = df.groupby(['receipt_id', 'feature_transaction_id']).apply(lambda x: x.sort_values(by='Success', ascending=False)).drop('Success', axis=1)
    
    # Dropping the unnecessary columns

    df = df[column_names]
    
    # Returning a dataframe 
    
    return df
    
    
    
    

In [40]:
df2 = preprocess('data_interview_test.csv')

In [41]:
df2

receipt_id    company_id  \
receipt_id feature_transaction_id                                   
10         000                    0              10  000:10000:10   
                                  1448           10  135:10000:10   
                                  1510           10  141:10000:10   
                                  1530           10  143:10000:10   
                                  1550           10  144:10000:10   
                                  1570           10  145:10000:10   
                                  1629           10  150:10000:10   
                                  1710           10  157:10000:10   
                                  1752           10  161:10000:10   
                                  1781           10  164:10000:10   
                                  1834           10  167:10000:10   
                                  1865           10  170:10000:10   
                                  1885           10  171:10000:10   
                                  1926           10  175:10000:10   
                                  1973           10  180:10000:10   
                                  1993           10  182:10000:10   
                                  2056           10  188:10000:10   
                                  2106           10  195:10000:10   
                                  2129           10  197:10000:10   
                                  2167           10  200:10000:10   
                                  2200           10  203:10000:10   
                                  2235           10  209:10000:10   
                                  2255           10  210:10000:10   
                                  2265           10  211:10000:10   
                                  2285           10  212:10000:10   
                                  1470           10  137:10000:10   
                                  1382           10  130:10000:10   
                                  42             10  004:10000:10   
                                  1355           10  128:10000:10   
                                  77             10  006:10000:10   
...                                             ...           ...   
50         367                    11978          50  221:50000:50   
                                  12032          50  226:50000:50   
           368                    12033          50  226:50000:50   
                                  10678          50  009:50000:50   
                                  10876          50  044:50000:50   
                                  10970          50  058:50000:50   
                                  11857          50  199:50000:50   
           369                    10679          50  009:50000:50   
                                  10877          50  044:50000:50   
           370                    11577          50  157:50000:50   
           376                    11444          50  132:50000:50   
                                  11638          50  168:50000:50   
           377                    11445          50  132:50000:50   
                                  11639          50  168:50000:50   
           378                    10812          50  033:50000:50   
           379                    10813          50  033:50000:50   
           407                    11134          50  089:50000:50   
                                  11078          50  079:50000:50   
           408                    11079          50  079:50000:50   
                                  11135          50  089:50000:50   
           409                    11080          50  079:50000:50   
                                  11136          50  089:50000:50   
           410                    11081          50  079:50000:50   
                                  11137          50  089:50000:50   
           411                    11082          50  079:50000:50   
                                  11138

# The algorithm is successful as we can see all the correct transactions for a receipt id and transaction id seem to be at the top of the list

## for example : For receipt id 50 and transaction id 368, there are two transactions associated but the correct one is seen at the top as the receipt id and transaction id both are matched correctly